### PR TITLE
DEV: Update API calls to ``datetime`` functionality

### PR DESCRIPTION
**Pull request by czgdp1807**
Continues - https://github.com/executablebooks/jupyter-cache/pull/114. #114 uses the latest APIs from `datetime` which works with Python >= 3.11. However, one needs to make it work with Python 3.9, Python 3.10 as well. For that one has to use `datetime.timezone.utc` and not `datetime.UTC` (available only in >= 3.11). See the review thread - https://github.com/executablebooks/jupyter-cache/pull/117#discussion_r1727041859

The API calls in `main` are depreciated. I have updated these calls to comply with the latest versions. If I use ``main`` to build SciPy docs with `myst_nb` execution mode set to ``cache`` then I get the following errors,

```zsh
reading sources... [  3%] reference/generated/scipy.cluster.hierarchy.DisjointSet.add .. reference/generated/scipy.fft.ifft
(Error in parallel process)
Traceback (most recent call last):
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/site-packages/sphinx/util/parallel.py", line 76, in _process
    ret = func(arg)
          ^^^^^^^^^
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 495, in read_process
    self.read_doc(docname, _cache=False)
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 536, in read_doc
    publisher.publish()
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/site-packages/docutils/core.py", line 234, in publish
    self.document = self.reader.read(self.source, self.parser,
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/site-packages/sphinx/io.py", line 106, in read
    self.parse()
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/site-packages/docutils/readers/__init__.py", line 76, in parse
    self.parser.parse(self.input, document)
  File "/Users/czgdp1807/Quansight/MyST-NB/myst_nb/sphinx_.py", line 152, in parse
    with create_client(
  File "/Users/czgdp1807/Quansight/MyST-NB/myst_nb/core/execute/base.py", line 79, in __enter__
    self.start_client()
  File "/Users/czgdp1807/Quansight/MyST-NB/myst_nb/core/execute/cache.py", line 36, in start_client
    _, self._notebook = cache.merge_match_into_notebook(self.notebook)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/czgdp1807/Quansight/jupyter-cache/jupyter_cache/cache/main.py", line 357, in merge_match_into_notebook
    cache_nb = self.get_cache_bundle(pk).nb
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/czgdp1807/Quansight/jupyter-cache/jupyter_cache/cache/main.py", line 296, in get_cache_bundle
    NbCacheRecord.touch(pk, self.db)
  File "/Users/czgdp1807/Quansight/jupyter-cache/jupyter_cache/cache/db.py", line 371, in touch
    record.accessed = datetime.utcnow()
                      ^^^^^^^^^^^^^^^^^
DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).

Sphinx parallel build error:
DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
make: *** [html-build] Error 2
``` 

With my PR I get the following,

```zsh
writing additional pages... search done
copying images... [  9%] ../build/plot_directive/reference/generated/scipy-interpolate-SmoothSphereBivariateSpline-__call__-1.pn
copying images... [100%] ../build/plot_directive/tutorial/stats/sampling_srou-1.png
dumping search index in English (code: en)... done
dumping object inventory... done
[jupyterlite-sphinx] Running JupyterLite build
[jupyterlite-sphinx] JupyterLite build done
build succeeded.

The HTML pages are in build/html.
```

### CODE REVIEW

**Comment by czgdp1807** [comment]
Seems like the newer API is applicable from Python 3.11 onwards.

**Comment by czgdp1807** [comment]
Requesting @chrisjsewell @choldgraf @agoose77 for review. Thanks. <3.

File: jupyter_cache/cache/db.py [hunk 1]
```diff
@@ -51,6 +52,12 @@ def get_version(path: Union[str, Path]) -> Optional[str]:
         return version_file.read_text().strip()
 
 
+def datetime_utcnow():
+    if sys.version_info.minor >= 11:
+        return lambda: datetime.datetime.now(datetime.UTC)
+    return datetime.datetime.utcnow
+
```
[the remark this answers is not in the document]
Comment (czgdp1807) [reply jupyter_cache/cache/db.py:59]: Actually `datetime.UTC` is available only in Python >= 3.11. Let me try with `datetime.datetime.now()`.

[the remark this answers is not in the document]
Comment (agoose77) [reply jupyter_cache/cache/db.py:59]: @czgdp1807 ah, try `datetime.timezone.UTC`

[the remark this answers is not in the document]
Comment (czgdp1807) [reply jupyter_cache/cache/db.py]: Addressed.

**Comment by czgdp1807** [comment]
There is a codecov error with Python 3.10 job.

**Comment by bsipocz** [comment]
This is a duplicate of #114. Either way, please let me know how I can help moving a release fwd as I run into these deprecations in a few projects, too.

**Comment by czgdp1807** [comment]
@bsipocz I noticed your PR as well. I think merging this PR (this one is compatible with Python 3.9, 3.10, 3.11 and 3.12) and releasing a new version of `jupyter-cache` should get things going. 

Requesting @agoose77 for reviewing and if possible merge this PR.

**Comment by bsipocz** [comment]
Well, the other would have been compatible too two weeks ago if CI was allowed to run on it 🤷‍♀️ 
And if you noticed it, it would have been a minimum courtesy to at least cross-reference an existing PR before starting a new one or at least once you notice it.

**Comment by czgdp1807** [comment]
> Well, the other would have been compatible too two weeks ago if CI was allowed to run on it 🤷‍♀️

Please check the description of this PR. I personally don't know why CI wasn't executed on your PR. Maintainers of this repository can comment on it.

> And if you noticed it, it would have been a minimum courtesy to at least cross-reference an existing PR before starting a new one or at least once you notice it.

Apologies. Referenced your PR here as well.

**Pull request merged**